### DISCUSSION TITLE
KNX and Scripts, KNX objects missing

### DISCUSSION THREAD

**Discussion by tJpADUNCcs** [arendst/Tasmota#21948] [Q&A] [answered]
Hi,

I try to build Tasmota for an ESP8266 including KNX and Smart Meter Interface features.
Therefore I modify user_config_override.h and am adding according to documentation:

```
#ifndef USE_SCRIPT
#define USE_SCRIPT
#endif
#ifndef USE_SML_M
#define USE_SML_M
#endif
#ifdef USE_RULES
#undef USE_RULES
#endif
#ifndef USE_KNX
#define USE_KNX         // Enable KNX IP Protocol Support (+9.4k code, +3k7 mem)
#endif
```

Build is working fine. But within KNX config I am missing KNX TX1 to KNX TX5 objects:
<img width="530" alt="image" src="https://github.com/user-attachments/assets/9b8f30bd-f4d0-4231-971d-dd089e14b814">

If I add KNX feature only (and not Script, SML and Rules) KNX TX1 TO KNX TX5 objects are available:
<img width="530" alt="image" src="https://github.com/user-attachments/assets/a1b9ee13-943d-41b0-8924-3e09700558e3">

What am I doing wrong, or how do I need to build this different?

Thx!

**Comment by HofeBY** [comment]
Maybe my site help´s to get it started.

https://github.com/HofeBY/Tasmota-Firmware

read, maybe translate Liesmich-zuerst.txt to your mother language.

tasmota.bin is a functional in German language.

user_config_override.h was used to compile this version.

Greetings from Oberfranken
HofeBY

**Reply by barbudor** [reply]
KNX TX1 to TX5 are only available with Rules
They are not supported with Scripting Language

**Reply by HofeBY** [reply]
Maybe if needed (KNX TX1 to TX5) bad new´s for @tJpADUNCcs
He can also try RULE-tasmota.bin, without SCRIPT.

**Reply by barbudor** [reply]
But he needs SML


May be @gemu2015 could investigate how to add support for KNX TX in Scripting-Language ?

**Reply by tJpADUNCcs** [reply]
Indeed, I need SML and KNX.

**Reply by gemu2015** [reply]
See scripting docs cmd knx()

**Reply by tJpADUNCcs** [reply]
scripting docs only mentioned this
> knx(code value) = sends a number value to KNX

Would use it in scripting like this 
`
if upsecs%tper==0 {
knx(26 sml[1])
}
`

But am still missing the objects in knx menu...

**Reply by gemu2015** [reply]
did you compile with USE_KNX?

**Reply by tJpADUNCcs** [reply]
yes, like this:

```
#ifndef USE_SCRIPT
#define USE_SCRIPT
#endif
#ifndef USE_SML_M
#define USE_SML_M
#endif
#ifdef USE_RULES
#undef USE_RULES
#endif
#ifndef USE_KNX
#define USE_KNX         // Enable KNX IP Protocol Support (+9.4k code, +3k7 mem)
#endif
```

**Comment by tJpADUNCcs** [comment]
Interesting wise I did find a tasmota.bin file with scripting and knx enabled and available KNX TX1 to KNX TX5 objects:
[Tasmota 13.1.zip](https://github.com/user-attachments/files/16575791/Tasmota.13.1.zip)

How to build this?

**Reply by barbudor** [reply]
Indeed interresting. 13.1 is from August 2023 but I made 4 PR on KNX in 2024, so I may have broken something.
I'll investigate this week but can't promise exactly when

To build 13.1 you need to pull the proper tag in git

**Reply by tJpADUNCcs** [reply]
Not sure...
I did download 13.0.0 and 13.1.0 fresh from repository and build with 

```
#ifndef USE_KNX
#define USE_KNX
#endif
#ifndef USE_SCRIPT
#define USE_SCRIPT
#endif
#ifndef USE_SML_M
#define USE_SML_M
#endif
#ifdef USE_RULES
#undef USE_RULES
#endif
```

and still the same. In KNX menu I do not have KNX TX1 to KNX TX5 objects for selection!
As soon as I do not build with script, TX1 to TX5 are available...

(I do use VS Code with PlatformIO on Mac OS)

**Comment by gemu2015** [comment]
i checked the source code and the script knx() cmd uses a standard call to send a knx sensor value also used in many other drivers.
i implemented this years ago for a knx user and at that time it worked.

**Reply by tJpADUNCcs** [reply]
knx() cmd is not the issue. I can not select any objects in knx menu as KNX TX1 to KNX TX5 objects are not available (see first screenshot) as soon as I build with scripts function enabled...

**Reply by gemu2015** [reply, marked as answer]
this version allows TX1 to TX5 also in scripts
[xdrv_11_knx.ino.zip](https://github.com/user-attachments/files/16579450/xdrv_11_knx.ino.zip)

**Reply by tJpADUNCcs** [reply]
Thanks, KNX options are back! Will need to double check if KNX communication is working, but this is looking good!

**Comment by Laudi111** [comment]
Hey,
is there also a fix for Tasmota 14.3.0.4 with RX/TX1 to RX/TX9 for using with scripts? I still get nothing in the dropdown menu.
And is it possible to use the different KNX datatypes with scripts? KnxTx_Float and KnxTx_Byte are new since 14.3.

**Reply by gemu2015** [reply]
i did a pr with the needed changes today.

**Comment by Laudi111** [comment]
Thank you, it's working now. 
I have another question, how kann i use received KNX Messages in Tasmota Scripts?
So whats the scripting name for KNXRX_VAL2?
14:01:13.251 KNX: Empfangen von 6/6/10 Befehl Schreiben: 16391.46 zu KNX RX 2
14:01:13.258 RSL: RESULT = {"Command":"Unknown","Input":"EVENT KNXRX_VAL2=16391.46"}
